nifi-extension-bundles/nifi-network-bundle/nifi-network-processors/src/main/java/org/apache/nifi/processors/network/pcap/PCAP.java

```diff
@@ -0,0 +1,438 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package org.apache.nifi.processors.network.pcap;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ListIterator;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * PCAP (named after libpcap / winpcap) is a popular format for saving
+ * network traffic grabbed by network sniffers. It is typically
+ * produced by tools like [tcpdump](<a href="https://www.tcpdump.org/">...</a>) or
+ * [Wireshark](<a href="https://www.wireshark.org/">...</a>).
+ *
+ * @see <a href=
+ *      "https://wiki.wireshark.org/Development/LibpcapFileFormat">Source</a>
+ */
+public class PCAP {
+    private ByteBufferInterface io;
+
+    public PCAP(ByteBufferInterface io) {
+        this(io, null, null);
+    }
+
+    public PCAP(ByteBufferInterface io, Object parent, PCAP root) {
+
+        this.parent = parent;
+        this.root = root == null ? this : root;
+        this.io = io;
+        read();
+    }
+
+    public PCAP(Header hdr, List<Packet> packets) {
+        this.hdr = hdr;
+        this.packets = packets;
+    }
+
+    public byte[] readBytesFull() {
+
+        int headerBufferSize = 20 + this.hdr().magicNumber().length;
+        ByteBuffer headerBuffer = ByteBuffer.allocate(headerBufferSize);
+        headerBuffer.order(ByteOrder.LITTLE_ENDIAN);
+
+        headerBuffer.put(this.hdr().magicNumber());
+        headerBuffer.put(this.readIntToNBytes(this.hdr().versionMajor(), 2, false));
+        headerBuffer.put(this.readIntToNBytes(this.hdr().versionMinor(), 2, false));
+        headerBuffer.put(this.readIntToNBytes(this.hdr().thiszone(), 4, false));
+        headerBuffer.put(this.readLongToNBytes(this.hdr().sigfigs(), 4, true));
+        headerBuffer.put(this.readLongToNBytes(this.hdr().snaplen(), 4, true));
+        headerBuffer.put(this.readLongToNBytes(this.hdr().network(), 4, true));
+
+        List<byte[]> packetByteArrays = new ArrayList<>();
+
+        ListIterator<Packet> packetsIterator = packets.listIterator();
+
+        int packetBufferSize = 0;
+
+        for (int loop = 0; loop < packets.size(); loop++) {
+            Packet currentPacket = packetsIterator.next();
+            ByteBuffer currentPacketBytes = ByteBuffer.allocate(16 + currentPacket.rawBody().length);
+            currentPacketBytes.put(readLongToNBytes(currentPacket.tsSec, 4, false));
+            currentPacketBytes.put(readLongToNBytes(currentPacket.tsUsec, 4, false));
+            currentPacketBytes.put(readLongToNBytes(currentPacket.inclLen, 4, false));
+            currentPacketBytes.put(readLongToNBytes(currentPacket.origLen, 4, false));
+            currentPacketBytes.put(currentPacket.rawBody());
+
+            packetByteArrays.add(currentPacketBytes.array());
+            packetBufferSize += 16 + currentPacket.rawBody().length;
+        }
```

Comment:
Sorry I did not notice this before.
```suggestion
        int packetBufferSize = 0;

        for (Packet currentPacket : packets) {
            ByteBuffer currentPacketBytes = ByteBuffer.allocate(16 + currentPacket.rawBody().length);
            currentPacketBytes.put(readLongToNBytes(currentPacket.tsSec, 4, false));
            currentPacketBytes.put(readLongToNBytes(currentPacket.tsUsec, 4, false));
            currentPacketBytes.put(readLongToNBytes(currentPacket.inclLen, 4, false));
            currentPacketBytes.put(readLongToNBytes(currentPacket.origLen, 4, false));
            currentPacketBytes.put(currentPacket.rawBody());

            packetByteArrays.add(currentPacketBytes.array());
            packetBufferSize += 16 + currentPacket.rawBody().length;
        }
```